src/canary.rs

```diff
@@ -186,3 +185,88 @@ impl Canary {
         }
     }
 }
+
+/// Write [`CANARY_VALUE`] to the stack.
+fn paint_stack(core: &mut Core, start: u32, mut end: u32) -> Result<(), probe_rs::Error> {
+    // make sure stack_end is properly aligned
+    if end % 4 != 0 {
+        end += end % 4;
+    }
+
+    // does the subroutine fit inside the stack?
+    let stack_size = (end - start) as usize;
+    assert!(
+        SUBROUTINE_LENGTH < stack_size,
+        "subroutine doesn't fit inside stack"
+    );
+
+    // write subroutine to RAM
+    // NOTE: add `SUBROUTINE_LENGTH` to `start`, to avoid the subroutine overwriting itself
+    core.write_8(start, &subroutine(start + SUBROUTINE_LENGTH as u32, end))?;
+
+    // store current PC and set PC to beginning of subroutine
+    let previous_pc = core.read_core_reg(PC)?;
+    core.write_core_reg(PC, start)?;
+
+    // execute the subroutine and wait for it to finish
+    core.run()?;
+    core.wait_for_core_halted(TIMEOUT).unwrap();
+
+    // overwrite subroutine
+    core.write_8(start, &[CANARY_VALUE; SUBROUTINE_LENGTH])?;
+
+    // reset PC to where it was before
+    core.write_core_reg(PC, previous_pc)?;
+
+    Ok(())
+}
+
+/// The length of the subroutine.
+const SUBROUTINE_LENGTH: usize = 28;
+
+/// Create a subroutine to paint [`CANARY_VALUE`] from `start` till `end`.
+///
+/// Both `start` and `end` need to be 4-byte-aligned.
+//
+// Roughly corresponds to following assembly:
+//
+// 00000108 <start>:
+//  108:   4803        ldr r0, [pc, #12]   ; (118 <end+0x2>)
+//  10a:   4904        ldr r1, [pc, #16]   ; (11c <end+0x6>)
+//  10c:   4a04        ldr r2, [pc, #16]   ; (120 <end+0xa>)
+//
+// 0000010e <loop>:
+//  10e:   4281        cmp r1, r0
+//  110:   d001        beq.n   116 <end>
+//  112:   c004        stmia   r0!, {r2}
+//  114:   e7fb        b.n 10e <loop>
+//
+// 00000116 <end>:
+//  116:   be00        bkpt    0x0000
+//  118:   20000100    .word   0x20000100  ; start
+//  11c:   20000200    .word   0x20000200  ; end
+//  120:   aaaaaaaa    .word   0xaaaaaaaa  ; pattern
+fn subroutine(start: u32, end: u32) -> [u8; SUBROUTINE_LENGTH] {
+    assert!(start < end, "start needs to be smaller than end address");
+    assert_eq!(start % 4, 0, "`start` needs to be 4-byte-aligned");
+    assert_eq!(end % 4, 0, "`end` needs to be 4-byte-aligned");
+
+    // convert start and end address to bytes
+    let [s1, s2, s3, s4] = start.to_le_bytes();
+    let [e1, e2, e3, e4] = end.to_le_bytes();
+
+    const CV: u8 = CANARY_VALUE;
+    [
+        0x03, 0x48, // ldr
```

Comment:
I'd paste the full instruction in the comment, so `ldd r0, [pc, #12]`.

Reply:
Do you think we need then need it also above the function? Or is inside enough?

Reply:
I like the one above as it is because it's a full objdump. But in this arrays there are three LDR instructions in a row, and I think we just need a little bit more context to see which one is which.